False-positive: public declarations in a dependency reported as redundant public accessibility

## Description

Periphery 3.5.0 incorrectly reports `public` declarations inside a Swift package as having redundant public accessibility, even though they **are used from outside the package** by the main app target.

## Issue

When a Swift package exposes `public` types and functions, and the main app target imports and uses them, Periphery still warns that the public accessibility is redundant ("not used outside of SampleDS").

## Example

**Repository:** https://github.com/AllDmeat/PeripheryIssueSample
**Commit hash:** `b90205a0f5da719219d1d249c7d70e3ac44c0ae6`

### Package: `SampleDS/Sources/SampleDS/SampleDS.swift`

```swift
import SwiftUI

public struct ClearButtonStyle: ButtonStyle {
    public func makeBody(configuration: Self.Configuration) -> some View {
        configuration.label
    }
}

extension ButtonStyle where Self == ClearButtonStyle {
    /// Remove SwiftUI default button styles with highlight and other styles.
    public static var clear: Self {
        ClearButtonStyle()
    }
}
```

### Main App Usage: `MyProject/ClearButtonView.swift`

```swift
import SwiftUI
import SampleDS

struct ClearButtonView: View {
    var body: some View {
        Button("Clear") { }
            .buttonStyle(.clear)
    }
}
```

The `ClearButtonView` is then embedded in `ViewController` via `UIHostingController`, so the entire chain is actively used.

### Periphery Output

```
/Users/alldmeat/Developer/PeripheryIssueSample/MyProject/SampleDS/Sources/SampleDS/SampleDS.swift:3:15: warning: Redundant public accessibility for struct 'ClearButtonStyle' (not used outside of SampleDS)
/Users/alldmeat/Developer/PeripheryIssueSample/MyProject/SampleDS/Sources/SampleDS/SampleDS.swift:4:17: warning: Redundant public accessibility for function 'makeBody(configuration:)' (not used outside of SampleDS)
```

## Expected Behavior

`ClearButtonStyle` and `makeBody(configuration:)` should **not** be flagged as having redundant public accessibility, since they are used by `MyProject` target which imports the `SampleDS` module.

## Environment

- Periphery version: 3.5.0
- Xcode version: 26.2 (Build version 17C52)
- Swift version: 6.2.3

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@
 
 ##### Bug Fixes
 
+- Fix false positive redundant public accessibility for types used in same-type generic requirements on extensions, e.g. `extension SomeProtocol where Self == MyType`.
 - Exclude declarations ignored by a file-level `// periphery:ignore:all` comment from superfluous ignore commend detection.
 
 ## 3.5.0 (2026-02-11)

--- a/Sources/SourceGraph/Mutators/RedundantExplicitPublicAccessibilityMarker.swift
+++ b/Sources/SourceGraph/Mutators/RedundantExplicitPublicAccessibilityMarker.swift
@@ -124,7 +124,21 @@ final class RedundantExplicitPublicAccessibilityMarker: SourceGraphMutator {
                 return nil
             }
 
-        return referenceDecls.contains { $0.accessibility.isAccessibleCrossModule }
+        return referenceDecls.contains {
+            if $0.accessibility.isAccessibleCrossModule {
+                return true
+            }
+
+            // Extensions may not have explicit public accessibility themselves, but their members
+            // can be public. Types used in an extension's generic constraints (e.g.,
+            // `extension SomeProtocol where Self == SomeType`) are effectively exposed publicly
+            // if the extension contains any public members.
+            if $0.kind.isExtensionKind {
+                return $0.declarations.contains { $0.accessibility.isAccessibleCrossModule }
+            }
+
+            return false
+        }
     }
 
     /// A public protocol that is not directly referenced cross-module may still be exposed by a public member declared

--- a/Sources/SyntaxAnalysis/DeclarationSyntaxVisitor.swift
+++ b/Sources/SyntaxAnalysis/DeclarationSyntaxVisitor.swift
@@ -482,6 +482,10 @@ public final class DeclarationSyntaxVisitor: PeripherySyntaxVisitor {
         return clause.requirements.reduce(into: .init()) { result, requirement in
             if let conformanceRequirementType = requirement.requirement.as(ConformanceRequirementSyntax.self) {
                 result.formUnion(typeSyntaxInspector.typeLocations(for: conformanceRequirementType.rightType))
+            } else if let sameTypeRequirement = requirement.requirement.as(SameTypeRequirementSyntax.self) {
+                if case let .type(rightType) = sameTypeRequirement.rightType {
+                    result.formUnion(typeSyntaxInspector.typeLocations(for: rightType))
+                }
             }
         }
     }

--- a/Tests/AccessibilityTests/AccessibilityProject/Sources/MainTarget/main.swift
+++ b/Tests/AccessibilityTests/AccessibilityProject/Sources/MainTarget/main.swift
@@ -79,3 +79,6 @@ inlinableFunction()
 // Associated types
 _ = PublicInheritedAssociatedTypeClass().items
 _ = PublicInheritedAssociatedTypeDefaultTypeClass().items
+
+// Extension with same-type generic requirement
+takeExtensionSameTypeGenericRequirement(.defaultInstance)

--- a/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsExtensionSameTypeGenericRequirement.swift
+++ b/Tests/AccessibilityTests/AccessibilityProject/Sources/TargetA/PublicTypeUsedAsExtensionSameTypeGenericRequirement.swift
@@ -1,0 +1,15 @@
+public protocol PublicTypeUsedAsExtensionSameTypeGenericRequirement_Protocol {}
+
+public struct PublicTypeUsedAsExtensionSameTypeGenericRequirement: PublicTypeUsedAsExtensionSameTypeGenericRequirement_Protocol {
+    public init() {}
+}
+
+extension PublicTypeUsedAsExtensionSameTypeGenericRequirement_Protocol where Self == PublicTypeUsedAsExtensionSameTypeGenericRequirement {
+    public static var defaultInstance: Self {
+        PublicTypeUsedAsExtensionSameTypeGenericRequirement()
+    }
+}
+
+public func takeExtensionSameTypeGenericRequirement<T: PublicTypeUsedAsExtensionSameTypeGenericRequirement_Protocol>(_ value: T) {
+    _ = value
+}

--- a/Tests/AccessibilityTests/RedundantPublicAccessibilityTest.swift
+++ b/Tests/AccessibilityTests/RedundantPublicAccessibilityTest.swift
@@ -331,6 +331,12 @@ final class RedundantPublicAccessibilityTest: SPMSourceGraphTestCase {
         assertNotRedundantPublicAccessibility(.protocol("PublicInheritedAssociatedTypeDefaultType"))
     }
 
+    func testPublicTypeUsedAsExtensionSameTypeGenericRequirement() {
+        index()
+
+        assertNotRedundantPublicAccessibility(.struct("PublicTypeUsedAsExtensionSameTypeGenericRequirement"))
+    }
+
     func testPublicComparableOperatorFunction() {
         index()
 

